akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -1372,19 +1370,42 @@ trait FlowOps[+Out, +Mat] {
    *
    * '''Cancels when''' downstream cancels
    */
-  def merge[U >: Out](that: Graph[SourceShape[U], _]): Repr[U] =
+  def merge[U >: Out, M](that: Graph[SourceShape[U], M]): Repr[U] =
     via(mergeGraph(that))
 
   protected def mergeGraph[U >: Out, M](that: Graph[SourceShape[U], M]): Graph[FlowShape[Out @uncheckedVariance, U], M] =
     GraphDSL.create(that) { implicit b ⇒
       r ⇒
-        import GraphDSL.Implicits._
         val merge = b.add(Merge[U](2))
         r ~> merge.in(1)
         FlowShape(merge.in(0), merge.out)
     }
 
   /**
+   * Merge the given [[Source]] to this [[Flow]], taking elements as they arrive from input streams,
+   * picking always the smallest of the available elements (waiting for one element from each side
```

Comment:
yeah, the point was that the intuition leads to the wrong conclusion, therefore it needs extra communication from our side.